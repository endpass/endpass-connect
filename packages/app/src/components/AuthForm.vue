<template>
  <v-frame :loading="!inited"> 
    <form data-test="auth-form" @submit.prevent="emitSubmit">
      <form-field>
        <message data-test="form-message">{{ message }}</message>
      </form-field>
      <form-field v-if="error">
        <message :error="true" data-test="error-message">{{ error }}</message>
      </form-field>
      <form-field v-if="showEmail">
        <v-input
          v-model="email"
          :invalid="email.length > 0 && !isEmailValid"
          :autofocus="true"
          name="email"
          placeholder="Enter your email..."
          data-test="email-input"
        />
      </form-field>
      <form-controls>
        <v-button
          v-if="showEmail"
          :disabled="!isEmailValid || loading"
          :submit="true"
          type="primary"
          data-test="submit-button"
        >{{ primaryButtonLabel }}</v-button>
        <v-button @click="emitCancel" data-test="cancel-button">Close</v-button>
      </form-controls>
    </form>
  </v-frame>  
</template>

<script>
import VFrame from './VFrame.vue';
import VInput from './VInput.vue';
import VButton from './VButton.vue';
import Message from './Message.vue';
import FormField from './FormField.vue';
import FormControls from './FormControls.vue';

export default {
  name: 'AuthForm',

  props: {
    inited: {
      type: Boolean,
      default: false,
    },

    loading: {
      type: Boolean,
      default: false,
    },

    showEmail: {
      type: Boolean,
      default: true,
    },

    message: {
      type: String,
      required: true,
    },

    error: {
      type: String,
      default: null,
    },
  },

  data: () => ({
    email: '',
  }),

  computed: {
    primaryButtonLabel() {
      return !this.loading ? 'Log in' : 'Loading...';
    },

    isEmailValid() {
      return (
        this.email && /[a-zA-Z._\-0-9]+@[a-z0-9]+\.[a-z]{2,}/g.test(this.email)
      );
    },
  },

  methods: {
    emitSubmit() {
      if (this.isEmailValid) {
        this.$emit('submit', this.email);
      }
    },

    emitCancel() {
      this.$emit('cancel');
    },
  },

  components: {
    VFrame,
    VButton,
    VInput,
    Message,
    FormField,
    FormControls,
  },
};
</script>
