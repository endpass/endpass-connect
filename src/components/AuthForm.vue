<template>
  <form class="auth-form" @submit.prevent="emitSubmit">
    <header class="auth-form__header">
      <img class="auth-form__logo" src="../assets/logo.png" alt="Endpass">
      Connect
    </header>
    <div class="auth-form__body">
      <section class="auth-form__message">
        <message>{{ message }}</message>
      </section>
      <section v-if="error" class="auth-form__message">
        <message :error="true">{{ error }}</message>
      </section>
      <section v-if="showEmail" class="auth-form__field">
        <v-input
          v-model="email"
          :invalid="email.length > 0 && !isEmailValid"
          :autofocus="true"
          name="email"
          placeholder="Enter your email..."
        />
      </section>
      <section class="auth-form__controls">
        <v-button
          v-if="showEmail"
          :disabled="!isEmailValid || loading"
          :submit="true"
          type="primary"
        >{{ primaryButtonLabel }}</v-button>
        <v-button @click="emitCancel">Close</v-button>
      </section>
    </div>
  </form>
</template>

<script>
import VInput from '@/components/VInput';
import VButton from '@/components/VButton';
import Message from '@/components/Message';

export default {
  name: 'AuthForm',

  props: {
    loading: {
      type: Boolean,
      default: false,
    },

    showEmail: {
      type: Boolean,
      default: true,
    },

    message: {
      type: String,
      required: true,
    },

    error: {
      type: String,
      default: null,
    },
  },

  data: () => ({
    email: '',
  }),

  computed: {
    primaryButtonLabel() {
      return !this.loading ? 'Log in' : 'Loading...';
    },

    isEmailValid() {
      return (
        this.email && /[a-zA-Z._\-0-9]+@[a-z0-9]+\.[a-z]{2,}/g.test(this.email)
      );
    },
  },

  methods: {
    emitSubmit() {
      this.$emit('submit', this.email);
    },

    emitCancel() {
      this.$emit('cancel');
    },
  },

  components: {
    VButton,
    VInput,
    Message,
  },
};
</script>

<style lang="postcss">
@keyframes slideIn {
  from {
    transform: translateY(15px);
    opacity: 0;
  }
}

.auth-form {
  overflow: hidden;
  max-width: 360px;
  margin: 50px auto;
  border-radius: 4px;
  box-shadow: 0 5px 10px 1px rgba(0, 0, 0, 0.15);
  background-color: #fff;
  animation: slideIn 0.75s;
}

.auth-form__header {
  display: flex;
  align-items: center;
  padding: 15px;
  font-size: 1.5em;
  background-color: #4b0472;
  color: #fff;
}

.auth-form__logo {
  flex: 0 0 auto;
  width: 32px;
  height: 32px;
  margin-right: 12.5px;
}

.auth-form__body {
  padding: 30px 15px 15px;
  font-size: 1em;
}

.auth-form__field {
  margin-bottom: 15px;
}

.auth-form__message {
  margin-bottom: 20px;
}

.auth-form__controls {
  display: flex;
  align-items: center;
  margin-top: 30px;

  & > button {
    flex: 0 0 auto;

    &:last-child {
      margin-left: auto;
    }
  }
}
</style>
