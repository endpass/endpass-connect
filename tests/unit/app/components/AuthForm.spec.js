import { shallowMount, mount } from '@vue/test-utils';
import AuthForm from '@@/app/src/components/AuthForm.vue';

describe('AuthForm', () => {
  describe('render', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(AuthForm, {
        propsData: {
          inited: true,
        },
      });
    });

    it('should correctly render AuthForm component', () => {
      expect(wrapper.name()).toBe('AuthForm');
      expect(wrapper.find('[data-test=email-input]').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render error', () => {
      wrapper = shallowMount(AuthForm, {
        propsData: {
          message: 'foo',
          error: 'bar',
        },
      });

      expect(wrapper.findAll('[data-test=error-message]').exists()).toBe(true);
    });

    it('should change submit button text if loading and make it disabled', () => {
      wrapper = shallowMount(AuthForm, {
        propsData: {
          message: 'foo',
          loading: true,
        },
      });

      const submitButton = wrapper.find('[data-test=submit-button]');

      expect(wrapper.html()).toMatchSnapshot();
      expect(submitButton.text()).toBe('Loading...');
      expect(submitButton.attributes().disabled).toBe('true');
    });
  });

  describe('behavior', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(AuthForm, {
        propsData: {
          message: 'foo',
          inited: true,
        },
      });
    });

    it('should not allow submit form if email is empty', () => {
      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toBe(undefined);
    });

    it('should not allow submit form if email is empty', () => {
      wrapper.setData({
        email: 'foo@bar',
      });

      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toBe(undefined);
    });

    it('should allow submit of email is valid', () => {
      wrapper.setData({
        email: 'foo@bar.baz',
      });

      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toEqual([['foo@bar.baz']]);
    });

    it('should cancel auth on cancel button press', () => {
      wrapper.find('[data-test=cancel-button]').trigger('click');

      expect(wrapper.emitted().cancel).toEqual([[]]);
    });
  });
});
