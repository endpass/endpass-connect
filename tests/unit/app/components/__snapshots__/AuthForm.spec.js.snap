// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm render should change submit button text if loading and make it disabled 1`] = `
<v-frame-stub loading="true">
  <form data-test="auth-form">
    <form-field-stub>
      <message-stub data-test="form-message">foo</message-stub>
    </form-field-stub>
    <!---->
    <form-field-stub>
      <v-input-stub placeholder="Enter your email..." name="email" value="" autofocus="true" data-test="email-input"></v-input-stub>
    </form-field-stub>
    <form-controls-stub>
      <v-button-stub disabled="true" type="primary" submit="true" data-test="submit-button">Loading...</v-button-stub>
      <v-button-stub data-test="cancel-button">Close</v-button-stub>
    </form-controls-stub>
  </form>
</v-frame-stub>
`;

exports[`AuthForm render should correctly render AuthForm component 1`] = `
<v-frame-stub>
  <form data-test="auth-form">
    <form-field-stub>
      <message-stub data-test="form-message">foo</message-stub>
    </form-field-stub>
    <!---->
    <form-field-stub>
      <v-input-stub placeholder="Enter your email..." name="email" value="" autofocus="true" data-test="email-input"></v-input-stub>
    </form-field-stub>
    <form-controls-stub>
      <v-button-stub disabled="true" type="primary" submit="true" data-test="submit-button">Log in</v-button-stub>
      <v-button-stub data-test="cancel-button">Close</v-button-stub>
    </form-controls-stub>
  </form>
</v-frame-stub>
`;

exports[`AuthForm render should not render form if loading passed as true 1`] = `
<div class="frame">
  <header class="frame__header"><img src="../assets/logo.png" alt="Endpass" class="frame__logo">
    Connect
  </header>
  <div class="frame__body">
    <div class="loading-screen">
      <div class="loading-screen__spinner">
        <div class="spinner"></div>
      </div>
    </div>
  </div>
</div>
`;
